contracts/voting/Voting.sol

```diff
@@ -0,0 +1,342 @@
+        // solhint-disable-next-line not-rely-on-time
+        return (block.timestamp - _regions[regionId].startVotingTimestamp) /
+            (VOTING_DURATION + GOVERNANCE_DURATION + CLAIMING_DURATION);
+    }
+
+    function _isVotingPeriod(uint256 regionId) internal view returns(bool ok) {
+        // solhint-disable-next-line not-rely-on-time
+        ok = ((block.timestamp - _regions[regionId].startVotingTimestamp) %
+            (VOTING_DURATION + GOVERNANCE_DURATION + CLAIMING_DURATION)) < VOTING_DURATION;
+    }
+
+    function _isGoverningPeriod(uint256 regionId) internal view returns(bool) {
+        /* solhint-disable not-rely-on-time */
+        return 
+            ((block.timestamp - _regions[regionId].startVotingTimestamp) %
+                (VOTING_DURATION + GOVERNANCE_DURATION + CLAIMING_DURATION)) > VOTING_DURATION &&
+            ((block.timestamp - _regions[regionId].startVotingTimestamp) %
+                (VOTING_DURATION + GOVERNANCE_DURATION + CLAIMING_DURATION)) < (VOTING_DURATION + GOVERNANCE_DURATION);
+        /* solhint-enable not-rely-on-time */
+    }
+
+    function _isRewardPeriod(uint256 regionId) internal view returns(bool) {
+        // solhint-disable-next-line not-rely-on-time
+        return ((block.timestamp - _regions[regionId].startVotingTimestamp) %
+            (VOTING_DURATION + GOVERNANCE_DURATION + CLAIMING_DURATION)) > (VOTING_DURATION + GOVERNANCE_DURATION);
+    }
+
+    function _calculateVotesPrice(
+        uint256 mayorId,
+        uint256 cityId,
+        uint256 votes
+    ) internal view returns(uint256) {
+        if (!_isVotingPeriod(_cities[cityId].regionId)) revert VotingErrors.IncorrectPeriod();
+        Nominee[] memory nominees = _cityToNominees[cityId][_seasonNumber(_cities[cityId].regionId) + 1];
+        if(
+            votes > ((_cities[cityId].population * votesPerCitizen / 1 ether) - _getBank(nominees))
+        ) revert VotingErrors.VotesBankExceeded();
+        return (votes * _cities[cityId].votePrice * _getVoteMultiplier(mayorId, cityId, msg.sender)) / 100;
+    }
+
+    function _getBuildingsDiscount(uint256 cityId, address account) internal view returns(uint256) {
+        if (_ownerToBuildings[account][cityId][Building.University] > 0) {
+            return VotingConstants.BUILDING_DISCOUNT_UNIVERSITY;
+        } else if (_ownerToBuildings[account][cityId][Building.Hospital] > 0) {
+            return VotingConstants.BUILDING_DISCOUNT_HOSPITAL;
+        } else {
+            return 0;
+        }
+    }
+
+    function _getVoteMultiplier(
+        uint256 nftId,
+        uint256 cityId,
+        address account
+    ) internal view returns(uint256) {
+        return 100 - mayor.getVoteDiscount(nftId) - _getBuildingsDiscount(cityId, account);
+    }
+
+    function _calculateWinner(
+        uint256 season, 
+        uint256 cityId
+    ) internal view returns(uint256) {
+        Nominee[] memory nominees = _cityToNominees[cityId][season];
```

Comment:
If nominees array will be too long memory usage will consume a lot of gas. Try to use direct access to storage

Reply:
Done